group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1053,6 +1055,13 @@ ConsumerGroup convertToConsumerGroup(ClassicGroup classicGroup, List<Coordinator
 
             throw new GroupIdNotFoundException("Cannot upgrade the classic group " + classicGroup.groupId() +
                 " to consumer group because the embedded consumer protocol is malformed.");
+        } catch (UnsupportedVersionException e) {
+            log.warn("Cannot upgrade the classic group " + classicGroup.groupId() +
+                " to consumer group: " + e.getMessage() + ".", e);
+
+            throw new GroupIdNotFoundException("Cannot upgrade the classic group " + classicGroup.groupId() +
+                " to consumer group because a custom assignor with userData is in use. " +
```

Comment:
I wonder if we could make `userData` clearer. Most of the users won't understand it. Should we perhaps say `... because an unsupported custom assignor is in use. Please refer to the document or switch to ...`.

Reply:
To clarify, which document should I direct users to?

Reply:
Sorry, I meant `documentation`, which implies the Kafka doc.

Reply:
Reworded the error to `Cannot upgrade classic group %s to consumer group because an unsupported custom assignor is in use. Please refer to the documentation or switch to a default assignor before re-attempting the upgrade.`